consensus/src/validation/step.rs

```diff
@@ -54,32 +54,33 @@ impl<T: Operations + 'static> ValidationStep<T> {
         inbound: AsyncQueue<Message>,
         executor: Arc<Mutex<T>>,
     ) {
-        let hash = candidate.header().hash;
+        let header = candidate.header();
 
         // A Validation step with empty/default Block produces a Nil Vote
-        if hash == [0u8; 32] {
+        if header.hash == [0u8; 32] {
```

Comment:
this is not necessary. you can just check header+vst if hash!=0 and then call `cast_vote` with `header.hash`
ie:
if header.hash != [0u8; 32]: 
  check header
  check vst
  
cast_vote(header.hash)

Just saying, feel free to keep it this way if you prefer

Reply:
Current impl highlights the logic of voting NiL if the provided block is empty